### ISSUE TITLE
交叉验证

### ISSUE THREAD

**Issue by geek-APTX4869** [shijun18/mmTumorSeg#2]
亲爱的作者：
  表2说在单独的测试集上比较，但又说在训练集上交叉验证，究竟怎么回事呢？

**Comment by shijun18** [comment]
As shown in the figure below, during the experiment, we first divided the whole dataset into a training set and an independent test set with a ratio of 8:2. Then, we further divided the training set into 5-fold for cross-validation. For each fold of training, we use an early stopping strategy to select the best weights based on the model's performance on the current validation set. However, due to the small size of the dataset, the distribution of the validation set and the whole dataset are likely to be inconsistent, so we believe that the performance of the validation set cannot accurately reflect the ability of the model. Therefore, to more objectively evaluate the accuracy and robustness of the model, different from traditional cross-validation, we conducted a final performance evaluation on the fixed independent test set. The results reported in the paper are the average of the models trained with different folds on the test set.

![cross-validation](https://github.com/shijun18/H-DenseFormer/assets/46990068/fcbcac5e-733c-4411-922c-366d5a233615)

**Comment by geek-APTX4869** [comment]
Thanks for your kind help.I wrote a function, but it doesn't seem to be able to evaluate the 2D results correctly.
```
def eval_2d():
    # test data
    data_path = r'E:\PI-CAI22\test_2d_seg'
    test_transformer = transforms.Compose([
                MRNormalize(),
                To_Tensor(num_class=2)
            ])
    net=HDenseFormer_2D_32(in_channels=3, n_cls=2, image_size=(384,384), transformer_depth=24)
    pathlist = glob.glob(os.path.join(data_path,'*.hdf5'))
    all_JI,all_Dice,all_95=[],[],[]

    for i in range(1,6):
        ck_path='ckpt/PI-CAI22/2d_seg/fold{}/'.format(i)
        weight=os.listdir(ck_path)[-1]
        print(weight)
        # get weight
        checkpoint = torch.load(ck_path+weight)
        net.load_state_dict(checkpoint['state_dict'],strict=False)
        net = net.cuda()
        net.eval()
        JI,Dice,h95=[],[],[]
        for path in pathlist:
            image = hdf5_reader(path, 'ct')
            label = hdf5_reader(path, 'seg')
            sample = {'image': image, 'label': label}
            sample = test_transformer(sample)
            
            image = sample['image'].unsqueeze(0).cuda()

            with autocast(False):
                output = net(image)
                if isinstance(output,list):
                    output = output[0]
            # measure run dice
            output = torch.argmax(torch.softmax(
                output, dim=1), 1).detach().cpu().numpy().squeeze()  #N*H*W
            result = cal_score(output,label)
            JI.append(result['Jaccard'])
            Dice.append(result['Dice'])
            h95.append(result['HausdorffDistance95'])
            
        all_JI.append(np.mean(JI))
        all_Dice.append(np.mean(Dice))
        all_95.append(np.mean(h95))
        
    print('Dice均值：',np.mean(all_Dice),'标准差：',np.std(all_Dice))
    print('HD95均值：',np.mean(all_95),'标准差：',np.std(all_95))
    print('JI均值：',np.mean(all_JI),'标准差：',np.std(all_JI))
```
